packages/compass-crud/src/components/change-view/change-view.tsx

```diff
@@ -0,0 +1,587 @@
+  paddingLeft: 0,
+});
+
+function ChangeObject({
+  obj,
+  isOpen,
+  isRoot,
+}: {
+  obj: ObjectBranch;
+  isOpen: boolean;
+  isRoot?: boolean;
+}) {
+  // A sample object / sub-document. ie. not an array and not a leaf.
+  if (isOpen) {
+    return (
+      <div className={cx(changeObjectStyles, isRoot && rootChangeObjectStyles)}>
+        {obj.properties.map((property) => {
+          const key = getObjectKey(property);
+          return <ChangeObjectProperty key={key} property={property} />;
+        })}
+      </div>
+    );
+  }
+
+  return null;
+}
+
+function getLeftClassName(obj: UnifiedBranch, darkMode?: boolean) {
+  const addedClass = darkMode ? addedStylesDark : addedStylesLight;
+  const removedClass = darkMode ? removedStylesDark : removedStylesLight;
+
+  if (obj.implicitChangeType === 'removed') {
+    return removedClass;
+  }
+
+  if (obj.implicitChangeType === 'added') {
+    return addedClass;
+  }
+
+  if (obj.changeType === 'unchanged') {
+    return undefined;
+  }
+
+  if (obj.changeType === 'removed') {
+    return removedClass;
+  }
+
+  return obj.changeType === 'changed' ? removedClass : addedClass;
+}
+
+function getRightClassName(obj: UnifiedBranch, darkMode?: boolean) {
+  return darkMode ? addedStylesDark : addedStylesLight;
+}
+
+function getChangeType(obj: UnifiedBranch) {
+  if (['added', 'removed'].includes(obj.implicitChangeType)) {
+    // these are "sticky" as we descend
+    return obj.implicitChangeType;
+  }
+
+  return obj.changeType;
+}
+
+function lookupValue(path: ObjectPath, value: any): any {
+  const [head, ...rest] = path;
+  if (rest.length) {
+    return lookupValue(rest, value[head]);
+  }
+  return value[head];
+}
+
+const changeValueStyles = css({
+  display: 'inline-flex',
+  flexWrap: 'wrap',
+  columnGap: spacing[1], // when removed and added are next to each other
+  rowGap: '1px', // when removed & added wrapped
+});
+
+function ChangeLeaf({ obj }: { obj: UnifiedBranch }) {
+  // Anything that is not an object or array. This includes simple javascript
+  // values like strings, numbers, booleans and undefineds, but also dates or
+  // bson values.
+  const { left, right } = useContext(LeftRightContext) as LeftRightContextType;
+
+  const changeType = getChangeType(obj);
+  // We could be showing the left value (unchanged, removed), right value
+  // (added) or both (changed). Furthermore the left one could have no colour or
+  // it could be red and the right one is always green.
+  const includeLeft = ['unchanged', 'changed', 'removed'].includes(changeType);
+  const includeRight = ['changed', 'added'].includes(changeType);
+
+  const darkMode = useDarkMode();
+
+  const leftValue = includeLeft
+    ? lookupValue((obj.left as Branch).path, left)
+    : undefined;
+  const rightValue = includeRight
+    ? lookupValue((obj.right as Branch).path, right)
+    : undefined;
+
+  // TODO: BSONValue does not always show the bson type, so you can't spot bson type changes
```

Comment:
TODO is not really the right word, but I did want to call attention to this. Since we're using the same components as the Document component the BSON values behave the same. So a Double with the value of `1` or an Int32 with value `1` or a Long with the value `1` all look the same.

This might not be much of a problem. As long we keep the before/after values with their types they will be stringified as `new Double(1)`, `new Int32(1)` and `new Long(1)` respectively before being diffed and therefore it will show up as changed. But the "red" and "green" values will then look identical.

I just bring this up because during tech design some people mentioned bson type changes.

Reply:
(I removed the TODO)